adr/ADR-41.md

```diff
@@ -0,0 +1,256 @@
+5. The server `leaf1` received from its Leafnode called `leaf1` with connection `lid:8`
+6. The server `leaf1` published the message to a Client with the connection name "NATS CLI Version development" over connection `cid:10`
+
+This is a set of 3 Trace messages holding 6 Trace Events in total.
+
+## Trace Message Formats
+
+The full detail of the trace message types are best found in the NATS Server source code in [server/msgtrace.go](https://github.com/nats-io/nats-server/blob/main/server/msgtrace.go), here we'll call out a few specifics about these messages.
+
+Given this sample message - a `MsgTraceEvent`:
+
+```json
+{
+  "server": {
+    "name": "n3-lon",
+    "host": "n3-lon.js.devco.net",
+    "id": "NCCPZOHDJ4KZQ35FU7EDFNPWXILA7U2VJAUWPG7IFDWUADDANNVOWKRV",
+    "cluster": "lon",
+    "domain": "hub",
+    "ver": "2.11.0-dev",
+    "tags": [
+      "lon"
+    ],
+    "jetstream": true,
+    "flags": 3,
+    "seq": 151861,
+    "time": "2024-02-22T13:18:48.400821739Z"
+  },
+  "request": {
+    "header": {
+      "Nats-Trace-Dest": [
+        "demo"
+      ]
+    },
+    "msgsize": 36
+  },
+  "hops": 1,
+  "events": [
+    {
+      "type": "in",
+      "ts": "2024-02-22T13:18:48.400595984Z",
+      "kind": 0,
+      "cid": 5390,
+      "name": "NATS CLI Version development",
+      "acc": "one",
+      "subj": "x"
+    },
+    {
+      "type": "eg",
+      "ts": "2024-02-22T13:18:48.40062456Z",
+      "kind": 0,
+      "cid": 5376,
+      "name": "NATS CLI Version development",
+      "sub": "x"
+    },
+    {
+      "type": "eg",
+      "ts": "2024-02-22T13:18:48.400649361Z",
+      "kind": 1,
+      "cid": 492,
+      "name": "n1-lon",
+      "hop": "1"
+    }
+  ]
+}
+```
+
+Lets have a quick look at some key fields:
+
+|Key|Notes|
+|`server`|This is a standard Server Info structure that you will see in many of our advisories, just indicates which server sent the event|
+|`request`|Details about the message being traced, more details about `Nats-Trace-Hop` below|
+|`hops`|How many remote destination (routes, gateways, Leafnodes) this server is sending the message to| 
+|`events`|A list of `MsgTraceEvents` that happened within the server related to this message, see below|
+
+### Sorting
+
+These events form a Directed Acyclic Graph that you can sort using the Hop information. The Server that sends the message to another Route, Gateway or Leafnode will indicate it will publish to a number of `hops` and each server that receives the message will have the `Nats-Trace-Hop` header set in its `request`.  
+
+Given `Server` -> `Gateway` -> `Router` -> `Leafnode` -> `Client` the first remote side of the Gateway would have `1.1` and the Leafnode might have `1.1.1`. However if there were multiple subscribers and multiple hops you might end up with `1.2.3`. On the sending side the Egress events have a `hop` key that would be `1` in the first case etc.
+
```

Comment:
In paraphrasing a bit, the way I would explain "hops" is this way:
The origin server - the server that receives the initial message with the `Nats-Trace-Dest` header - will have a `hops` count indicating to how many other servers (routes, leafs, gateways) it has forwarded the message to. This is not necessarily the total number of servers that the message will traverse.
Each time a server forwards a message to a remote server, it appends a number to its own `Nats-Trace-Hop` header value. Since the origin server does not have one, if it forwards a message to say a ROUTE, that remote server would receive the message with a new header `Nats-Trace-Hop` with the value of `1`, then the origin server forwards to a GATEWAY, and that server would receive the message with `Nats-Trace-Hop` value set to `2`.
Each of these servers in turn, if forwarding a message to another server, would add an incremental number to their existing `Nats-Trace-Hop` value, which would result in `1.1`, and `2.1`, etc..
Take a simple example of the origin server that has a LEAF server, which in turn as another LEAF server (daisy chained). If a message with tracing is forwarded, the trace message emitted from the origin server would have `hops` to 1 (since the origin server forwarded directly only to the first LEAF remote server). The trace emitted from the first LEAF would have a `Nats-Trace-Hop` header with value `1`. It would also have `hops` set to 1 (assuming there was interest in the last LEAF server). Finally, the last LEAF server would have a `Nats-Trace-Hop` of `1.1` and would have no `hops` field (since value would be 0).